enterprise/server/cmd/action_runner/main.go

```diff
@@ -0,0 +1,662 @@
+	user := os.Getenv(repoUserEnvVarName)
+	token := os.Getenv(repoTokenEnvVarName)
+	if user == "" && token == "" {
+		return *repoURL, nil
+	}
+	u, err := url.Parse(*repoURL)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse repo URL %q: %s", *repoURL, err)
+	}
+	u.User = url.UserPassword(user, token)
+	return u.String(), nil
+}
+
+func invocationURL(invocationID string) string {
+	urlPrefix := *besResultsURL
+	if !strings.HasSuffix(urlPrefix, "/") {
+		urlPrefix = urlPrefix + "/"
+	}
+	return urlPrefix + invocationID
+}
+
+func readConfig() (*workflowconf.BuildBuddyConfig, error) {
+	f, err := os.Open(actionsConfigPath)
+	if err != nil {
+		return nil, fmt.Errorf("open %q: %s", actionsConfigPath, err)
+	}
+	c, err := workflowconf.NewConfig(f)
+	if err != nil {
+		return nil, fmt.Errorf("read %q: %s", actionsConfigPath, err)
+	}
+	return c, nil
+}
+
+func matchesAnyTrigger(action *workflowconf.Action, event, branch string) bool {
+	if action.Triggers == nil {
+		return false
+	}
+	if pushCfg := action.Triggers.Push; pushCfg != nil && event == pushEventName {
+		return matchesAnyBranch(pushCfg.Branches, branch)
+	}
+
+	if prCfg := action.Triggers.PullRequest; prCfg != nil && event == pullRequestEventName {
+		return matchesAnyBranch(prCfg.Branches, branch)
+	}
+	return false
+}
+
+func matchesAnyBranch(branches []string, branch string) bool {
+	if branches == nil {
+		return false
+	}
+	for _, b := range branches {
+		if b == branch {
+			return true
+		}
+	}
+	return false
+}
+
+func runCommand(ctx context.Context, executable string, args []string, env map[string]string, outputSink io.Writer) error {
+	cmd := exec.CommandContext(ctx, executable, args...)
+	cmd.Stdout = outputSink
+	cmd.Stderr = outputSink
+	for k, v := range env {
+		cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", k, v))
+	}
+	return cmd.Run()
+}
+
+func getExitCode(err error) int {
+	if err == nil {
+		return 0
+	}
+	if exitErr, ok := err.(*exec.ExitError); ok {
+		return exitErr.ExitCode()
+	}
+	return noExitCode
+}
+
+func validateFlags() error {
+	errs := []string{}
+	for name, flag := range map[string]*string{
+		"repo_url":        repoURL,
+		"commit_sha":      commitSHA,
+		"trigger_branch":  triggerBranch,
+		"trigger_event":   triggerEvent,
+		"bes_backend":     besBackend,
+		"bes_results_url": besResultsURL,
+	} {
+		if *flag == "" {
+			errs = append(errs, fmt.Sprintf("--%s unset", name))
+		}
+	}
+	if len(errs) == 0 {
+		return nil
+	}
+	return fmt.Errorf(strings.Join(errs, "; "))
+}
+
+func actionDebugString(action *workflowconf.Action) string {
+	yamlBytes, err := yaml.Marshal(action)
+	if err != nil {
+		return fmt.Sprintf("<failed to marshal action: %s>", err)
+	}
+	return string(yamlBytes)
+}
+
+func newUUID() string {
+	id, err := uuid.NewRandom()
+	if err != nil {
+		log.Printf("failed to generate UUID")
+		os.Exit(retryableExitCode)
+	}
+	return id.String()
+}
+
+func dial(backend string) (*grpc.ClientConn, error) {
```

Comment:
Is there a reason not to use our internal grpc_client dial code here?

Reply:
Oh nice, I didn't know about that -- looks like it'll work perfectly. Thanks!